File: articles/design-evaluation-learning-outcome-measurement/cpp/evaluation_value_engine.cpp
#include <algorithm>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <sstream>
#include <stdexcept>
#include <string>
#include <vector>

struct Intervention {
    std::string name;
    double outcome_improvement;
    double burden_reduction;
    double equity_performance;
    double trust_improvement;
    double durability;
    double operational_cost;
    double residual_risk;
};

struct ScoredIntervention {
    std::string name;
    double value;
    double penalty;
};

std::vector<std::string> split_csv_line(const std::string& line) {
    std::vector<std::string> fields;
    std::stringstream ss(line);
    std::string item;
    while (std::getline(ss, item, ',')) {
        fields.push_back(item);
    }
    return fields;
}

double to_double(const std::string& value) {
    try {
        return std::stod(value);
    } catch (...) {
        throw std::runtime_error("Unable to parse numeric value: " + value);
    }
}

double penalty(const Intervention& x) {
    return 0.50 * x.operational_cost + 0.50 * x.residual_risk;
}

double evaluation_value(const Intervention& x) {
    return 0.24 * x.outcome_improvement +
           0.20 * x.burden_reduction +
           0.20 * x.equity_performance +
           0.16 * x.trust_improvement +
           0.14 * x.durability -
           0.06 * penalty(x);
}

std::vector<Intervention> read_portfolio(const std::string& path) {
    std::ifstream file(path);
    if (!file) {
        throw std::runtime_error("Unable to open input file: " + path);
    }

    std::vector<Intervention> interventions;
    std::string line;
    std::getline(file, line); // header

    while (std::getline(file, line)) {
        if (line.empty()) {
            continue;
        }

        auto fields = split_csv_line(line);
        if (fields.size() < 10) {
            throw std::runtime_error("Malformed CSV row: " + line);
        }

        interventions.push_back(
            Intervention{
                fields[0],
                to_double(fields[3]),
                to_double(fields[4]),
                to_double(fields[5]),
                to_double(fields[6]),
                to_double(fields[7]),
                to_double(fields[8]),
                to_double(fields[9])
            }
        );
    }

    return interventions;
}

int main(int argc, char* argv[]) {
    const std::string default_path = "../data/raw/evaluation_portfolio_raw.csv";
    const std::string input_path = argc > 1 ? argv[1] : default_path;

    try {
        auto interventions = read_portfolio(input_path);
        std::vector<ScoredIntervention> scored;

        for (const auto& intervention : interventions) {
            scored.push_back(
                ScoredIntervention{
                    intervention.name,
                    evaluation_value(intervention),
                    penalty(intervention)
                }
            );
        }

        std::sort(
            scored.begin(),
            scored.end(),
            [](const ScoredIntervention& a, const ScoredIntervention& b) {
                return a.value > b.value;
            }
        );

        std::cout << "rank,intervention,evaluation_value,penalty\n";
        for (std::size_t i = 0; i < scored.size(); ++i) {
            std::cout << (i + 1) << ","
                      << scored[i].name << ","
                      << std::fixed << std::setprecision(6)
                      << scored[i].value << ","
                      << scored[i].penalty << "\n";
        }
    } catch (const std::exception& ex) {
        std::cerr << "ERROR: " << ex.what() << "\n";
        return 1;
    }

    return 0;
}

File: articles/design-evaluation-learning-outcome-measurement/go/evaluation_value_engine.go
package main

import (
	"encoding/csv"
	"fmt"
	"log"
	"os"
	"sort"
	"strconv"
)

type Intervention struct {
	Name               string
	OutcomeImprovement float64
	BurdenReduction    float64
	EquityPerformance  float64
	TrustImprovement   float64
	Durability         float64
	OperationalCost    float64
	ResidualRisk       float64
	Penalty            float64
	EvaluationValue    float64
}

func parseFloat(value string) float64 {
	parsed, err := strconv.ParseFloat(value, 64)
	if err != nil {
		log.Fatalf("unable to parse numeric value %q: %v", value, err)
	}
	return parsed
}

func penalty(x Intervention) float64 {
	return 0.50*x.OperationalCost + 0.50*x.ResidualRisk
}

func evaluationValue(x Intervention) float64 {
	return 0.24*x.OutcomeImprovement +
		0.20*x.BurdenReduction +
		0.20*x.EquityPerformance +
		0.16*x.TrustImprovement +
		0.14*x.Durability -
		0.06*x.Penalty
}

func loadPortfolio(path string) []Intervention {
	file, err := os.Open(path)
	if err != nil {
		log.Fatalf("unable to open input file: %v", err)
	}
	defer file.Close()

	reader := csv.NewReader(file)
	rows, err := reader.ReadAll()
	if err != nil {
		log.Fatalf("unable to read CSV: %v", err)
	}

	var items []Intervention

	for i, row := range rows {
		if i == 0 {
			continue
		}
		if len(row) < 10 {
			log.Fatalf("malformed row: %v", row)
		}

		x := Intervention{
			Name:               row[0],
			OutcomeImprovement: parseFloat(row[3]),
			BurdenReduction:    parseFloat(row[4]),
			EquityPerformance:  parseFloat(row[5]),
			TrustImprovement:   parseFloat(row[6]),
			Durability:         parseFloat(row[7]),
			OperationalCost:    parseFloat(row[8]),
			ResidualRisk:       parseFloat(row[9]),
		}
		x.Penalty = penalty(x)
		x.EvaluationValue = evaluationValue(x)
		items = append(items, x)
	}

	return items
}

func main() {
	path := "../data/raw/evaluation_portfolio_raw.csv"
	if len(os.Args) > 1 {
		path = os.Args[1]
	}

	items := loadPortfolio(path)

	sort.Slice(items, func(i, j int) bool {
		return items[i].EvaluationValue > items[j].EvaluationValue
	})

	fmt.Println("rank,intervention,evaluation_value,penalty")
	for i, x := range items {
		fmt.Printf("%d,%s,%.6f,%.6f\n", i+1, x.Name, x.EvaluationValue, x.Penalty)
	}
}
